base/compiler/loops.jl

```diff
@@ -0,0 +1,350 @@
+                          iterrettype,
+                          nextid)
+    end
+    if !isassigned(src.slotnames, slotid)
+        @assert length(src.slotnames) == slotid - 1
+        # push!(src.slotnames, Symbol(Main.Base.string(slotname, "[", i, "]")))
+        push!(src.slotnames, slotname)
+        push!(src.slotflags, 0x02)  # 0x02 = assigned
+        push!(src.slottypes, iterrettype)
+    end
+    # %d === nothing
+    addstmt!(block,  # deliberately don't increment nextid
+             Expr(:call, GlobalRef(Core, :(===)), iterret, nothing),
+             Bool,
+             nextid)
+    # Base.not_int(%d)
+    addstmt!(block,
+             Expr(:call, GlobalRef(Main.Base, :not_int), SSAValue(nextid)),
+             Bool,
+             nextid+1)
+    # goto #x if not %y
+    nextid = addstmt!(block,
+                      Expr(:gotoifnot, SSAValue(nextid+1), exitid),
+                      Any,
+                      nextid+2)
+    if isa(iterrettype, Union)
+        itemstatetype = iterrettype.a === Nothing ? iterrettype.b : iterrettype.a
+        itemtype  = itemstatetype.parameters[1]
+        statetype = itemstatetype.parameters[2]
+    else
+        itemstatetype = itemtype = statetype = Any
+    end
+    item = nothing
+    if state === nothing
+        # item, state = iterret
+        enterid = nextid
+        iterretssa = SSAValue(nextid)
+        # Union-splitting via insertion of a TypedSlot
+        nextid = addstmt!(block,
+                          TypedSlot(iterret.id, itemstatetype),
+                          itemstatetype,
+                          nextid)
+        item = SSAValue(nextid)
+        nextid = addstmt!(block,
+                          Expr(:call, GlobalRef(Core, :getfield), iterretssa, 1),
+                          itemtype,
+                          nextid)
+        state = SSAValue(nextid)
+        nextid = addstmt!(block,
+                          Expr(:call, GlobalRef(Core, :getfield), iterretssa, 2),
+                          statetype,
+                          nextid)
+    else
+        # Jump back to top of loop
+        nextid = addstmt!(block,
+                          GotoNode(enterid),
+                          Any,
+                          nextid)
+        enterid = nothing
+    end
+    return block, enterid, item, itemtype, state, statetype, nextid
+end
+
+# Convert an iteration over a CartesianIndices into a set of nested loops
+# `src.code[id]` must contain the call to `iterate(::CartesianIndices)`
+function expand_cartesian_loop!(src::CodeInfo, id::Int)
+    @assert isempty(src.ssaflags)
+    @assert !isempty(src.ssavaluetypes)  # puzzlingly, src.inferred might still be false
+    println(src)
+    iterator, iteratortype, itervar, bodystart, bodystop, loopexit = extract_iterate(src, id)
+    @assert iteratortype <: Main.Base.CartesianIndices
+    # slotname = isa(iterator, Slot) ? Main.Base.string("#", Main.Base.String(src.slotnames[iterator.id])) : Main.Base.String(gensym())
```

Comment:
Would be nice to be able to give things nice names but currently this causes bootstrap problems.

Reply:
In general `Main.Base` might not be available during bootstrap?

Reply:
This only gets called when it is, see the check in `expand_cartesian_loops!(src::CodeInfo)`